File: app/src/main/java/com/example/tylerbwong/awaken/utilities/AnimatedRecyclerView.java
package com.example.tylerbwong.awaken.utilities;

import android.content.Context;
import androidx.recyclerview.widget.RecyclerView;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

/**
 * @author Connor Wong
 */
public class AnimatedRecyclerView extends RecyclerView {
    private boolean mScrollable;
    private boolean mIsAnimatable = true;

    public AnimatedRecyclerView(Context context) {
        this(context, null);
    }

    public AnimatedRecyclerView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public AnimatedRecyclerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        mScrollable = false;
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        return !mScrollable || super.dispatchTouchEvent(ev);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        if (mIsAnimatable) {
            for (int i = 0; i < getChildCount(); i++) {
                animate(getChildAt(i), i);

                if (i == getChildCount() - 1) {
                    getHandler().postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            mScrollable = true;
                        }
                    }, i * 100);
                }
            }
        }
        mIsAnimatable = true;
    }

    public void setIsAnimatable(boolean isAnimatable) {
        this.mIsAnimatable = isAnimatable;
    }

    private void animate(View view, final int pos) {
        view.animate().cancel();
        view.setTranslationY(100);
        view.setAlpha(0);
        view.animate().alpha(1.0f).translationY(0).setDuration(300).setStartDelay(pos * 100);
    }
}

File: app/src/main/java/io/awaken/ui/utils/AnimatedRecyclerView.kt
package io.awaken.ui.utils

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import androidx.recyclerview.widget.RecyclerView

/**
 * @author Connor Wong
 */
class AnimatedRecyclerView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyle: Int = 0) : RecyclerView(context, attrs, defStyle) {
    private var scrollable = false
    private var isAnimatable = true

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        return !scrollable || super.dispatchTouchEvent(ev)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        if (isAnimatable) {
            for (i in 0 until childCount) {
                animate(getChildAt(i), i)

                if (i == childCount - 1) {
                    handler.postDelayed({ scrollable = true }, (i * 100).toLong())
                }
            }
        }
        isAnimatable = true
    }

    fun setIsAnimatable(isAnimatable: Boolean) {
        this.isAnimatable = isAnimatable
    }

    private fun animate(view: View, pos: Int) {
        view.animate().cancel()
        view.translationY = 100f
        view.alpha = 0f
        view.animate().alpha(1.0f).translationY(0f).setDuration(300).startDelay = (pos * 100).toLong()
    }
}
